### PR TITLE
Move LocalAnalyzer to rustc_mir::ssa_analyze

### PR DESCRIPTION
**Pull request by bjorn3**
This is useful for other codegen backends as well.

### CODE REVIEW

**Comment by estebank** [comment]
It's gonna take me a while to get to this. @arielb1 can you step in?

File: src/librustc_mir/ssa_analyze.rs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+use rustc_data_structures::control_flow_graph::dominators::Dominators;
+use rustc_data_structures::indexed_vec::{Idx, IndexVec};
+use rustc::mir::{self, Mir, Location};
+use rustc::mir::visit::{Visitor, PlaceContext};
+use rustc::ty::{self, TyCtxt, Ty, TypeFoldable};
+use rustc::ty::subst::Substs;
+use rustc::ty::layout::{TyLayout, LayoutError};
+
+pub trait LocalAnalyzerCallbacks<'tcx> {
+    fn ty_ssa_allowed(&self, ty: Ty<'tcx>) -> bool;
```
Comment (eddyb) [inline src/librustc_mir/ssa_analyze.rs]: It would be more useful for this to take a `TyLayout`.

File: src/librustc_mir/ssa_analyze.rs [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    let mut analyzer = LocalAnalyzer::new(tcx, mir, param_substs, callbacks);
+
+    analyzer.visit_mir(mir);
+
+    for (index, ty) in mir.local_decls.iter().map(|l| l.ty).enumerate() {
+        let ty = analyzer.monomorphize(&ty);
+        debug!("local {} has type {:?}", index, ty);
+        if !analyzer.callbacks.ty_ssa_allowed(ty) {
+            analyzer.not_ssa(mir::Local::new(index));
+        }
```
Comment (eddyb) [inline src/librustc_mir/ssa_analyze.rs]: This part could arguably stay in `rustc_codegen_llvm`.

File: src/librustc_mir/ssa_analyze.rs [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+                // ZSTs don't require any actual memory access.
+                let elem_ty = base_ty.projection_ty(self.tcx, &proj.elem).to_ty(self.tcx);
+                let elem_ty = self.monomorphize(&elem_ty);
+                if self.layout_of(elem_ty).is_zst() {
+                    return;
+                }
+
+                if let mir::ProjectionElem::Field(..) = proj.elem {
+                    if self.callbacks.ty_ssa_allowed(base_ty.to_ty(self.tcx)) {
```
Comment (eddyb) [inline src/librustc_mir/ssa_analyze.rs]: Can you check that you need this `if`?

File: src/librustc_mir/ssa_analyze.rs [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        } else {
+            self.first_assignment[local] = location;
+        }
+    }
+
+    fn layout_of(&self, ty: Ty<'tcx>) -> TyLayout<'tcx> {
+        self.tcx.layout_of(ty::ParamEnv::reveal_all().and(ty))
+            .unwrap_or_else(|e| match e {
+                LayoutError::SizeOverflow(_) => self.tcx.sess.fatal(&e.to_string()),
+                _ => bug!("failed to get layout for `{}`: {}", ty, e)
```
Comment (eddyb) [inline src/librustc_mir/ssa_analyze.rs]: This is not in `rustc_codegen_llvm` so ideally we would treat a failure to get the error just like some arbitrary layout that we know nothing about.

**Comment by eddyb** [comment]
IMO this shouldn't live in `rustc_mir` unless it's somewhat more general (identifying variables that are SSA-like *without* any concerns for layout). If layout is involved I'd rather it be in the `rustc_codegen_ssa` crate proposed in #45274 (@irinagpopa will work on that, FWIW).

File: src/librustc_mir/ssa_analyze.rs (hunk 2 above)
Comment (bjorn3) [reply src/librustc_mir/ssa_analyze.rs]: The `ty_ssa_allowed` function is defined by the user of `non_ssa_locals`.

**Comment by bjorn3** [comment]
Got it

**Comment by eddyb** [comment]
Note that layout *may not* need to be directly involved (i.e. this pass could just check for multiple writes, borrows, non-SSA reads, those sorts of things), and `rustc_codegen_llvm` would just mark each non-immediate/pair local as non-SSA to handle the layout side of things.
But I haven't checked for sure - you could try https://github.com/rust-lang/rust/pull/52208#discussion_r202407695 and see if anything breaks.

**Pull request closed without merging**